[FLINK-10600] Provide End-to-end test cases for modern Kafka connectors

yanghua opened the pull request:
## What is the purpose of the change

*This pull request provides End-to-end test cases for modern Kafka connectors*


## Brief change log

  - *Provide End-to-end test cases for modern Kafka connectors*

## Verifying this change


This change added tests and can be verified as follows:

  - *Added end-to-end integration tests for modern kafka connector *

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

yanghua commented:
@pnowojski The error log shows my new shell script has permission problem : `test_streaming_kafka.sh: Permission denied`. Where can I give it execution permission?

zentol commented:
```git update-index --chmod=+x <file>```

pnowojski commented on flink-end-to-end-tests/test-scripts/test-data/modern-kafka-common.sh:
```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env bash
```
Please deduplicate the code here with `kafka-common.sh`

pnowojski commented on flink-end-to-end-tests/test-scripts/test-data/modern-kafka-common.sh:
```diff
@@ -0,0 +1,152 @@
+SCHEMA_REGISTRY_PORT=8082
+SCHEMA_REGISTRY_URL=http://localhost:${SCHEMA_REGISTRY_PORT}
+
+function setup_kafka_dist {
```
for example it looks like this method differs only with the `KAFKA_URL` "parameter"

pnowojski commented on flink-end-to-end-tests/test-scripts/test-data/modern-kafka-common.sh:
```diff
@@ -0,0 +1,152 @@
+  sed -i -e "s+^\(log\.dirs\s*=\s*\).*$+\1$TEST_DATA_DIR/kafka+" $KAFKA_DIR/config/server.properties
+}
+
+function setup_confluent_dist {
```
is this used anywhere?

yanghua replied:
it used in `test_confluent_schema_registry`

pnowojski replied:
I can not find this anywhere. Is it in the scope of this PR? If not, please move it to the appropriate PR/commit.

pnowojski replied:
Please deduplicate those two `kafka-common` files. Now this `setup_confluent_dist` is not only duplicated but also unused.

yanghua replied:
Yes, I know. My strategy is divided into two steps:

1) First ensure that the end-to-end test passes;
2) Refactor it, still use the original `kafka-common`, and finally there should be no `modern-kafka-common`

Currently I am working on the first problem (the wrong dependency version causes the class to not be found).

pnowojski commented on flink-end-to-end-tests/test-scripts/test-data/modern-kafka-common.sh:
```diff
@@ -0,0 +1,152 @@
+  $KAFKA_DIR/bin/kafka-server-stop.sh
+  $KAFKA_DIR/bin/zookeeper-server-stop.sh
+
+  PIDS=$(jps -vl | grep -i 'kafka\.Kafka' | grep java | grep -v grep | awk '{print $1}')
```
Why the script for kafka 0.10 didn't have this killing part? Should it have?

pnowojski replied:
Yes, you are right. I was looking at some outdated code.

pnowojski commented on flink-end-to-end-tests/test-scripts/test-data/modern-kafka-common.sh:
```diff
@@ -0,0 +1,152 @@
+    --consumer-property group.id=$3 2> /dev/null
+}
+
+function send_messages_to_kafka_avro {
```
Is this used anywhere?

pnowojski commented on flink-end-to-end-tests/test-scripts/test-data/modern-kafka-common.sh:
```diff
@@ -0,0 +1,152 @@
+    | cut -d " " -f 4
+}
+
+function start_confluent_schema_registry {
```
is this used anywhere?

pnowojski commented on flink-end-to-end-tests/test-scripts/test_streaming_kafka.sh:
```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env bash
```
Please deduplicate this as well.

pnowojski commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/kafka/KafkaExample.java:
```diff
@@ -0,0 +1,81 @@
+ */
+public class KafkaExample {
+
+	public static void main(String[] args) throws Exception {
```
lines from 43 to 60 also could be deduplicated.

pnowojski commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/kafka/KafkaExample.java:
```diff
@@ -0,0 +1,81 @@
+			new FlinkKafkaProducer<>(
+				parameterTool.getRequired("output-topic"),
+				new KafkaEventSchema(),
+				parameterTool.getProperties()));
```
I think it would be better to use `EXACTLY_ONCE` mode here

yanghua commented:
@pnowojski Using maven-shade-plugin, it's hard to type a jar that doesn't contain flink kafka 0.10 or old kafka client, which is why FLINK-10107 appears. Unless we adopt the maven profile. But this is not very suitable.

pnowojski commented:
@yanghua do you mean this end-to-end test is failing because of https://issues.apache.org/jira/browse/FLINK-10107 and it is blocking this PR?

yanghua commented:
No, I mean if we package a jar which contains old connector and new connector and multiple kafka client this pr's end to end test would fail with the same reason. But it seems shade plugin can not pick or exclude specific version of dependencies.

yanghua commented:
@pnowojski I solved the test problem and it is now ready to be reviewed again.

pnowojski commented on flink-examples/flink-examples-streaming/pom.xml:
```diff
@@ -538,6 +550,11 @@ under the License.
 						<configuration>
+							<artifactSet>
+								<excludes>
+									<exclude>org.apache.flink:flink-connector-kafka_${scala.binary.version}</exclude>
```
Maybe this one should exclude `0.8`, `0.9` and `0.11` as well for the sake of  consistency?

pnowojski replied:
Ok, right. We can only exclude the modern one and `0.11`? And we can still exclude everything from "modern" `KafkaExample`?

yanghua replied:
yes.

pnowojski commented on flink-examples/flink-examples-streaming/pom.xml:
```diff
@@ -568,6 +585,49 @@ under the License.
+								</transformer>
+							</transformers>
+							<finalName>KafkaExample</finalName>
+							<!-- <outputFile>Kafka.jar</outputFile> -->
```
Why do we need this comment?

yanghua replied:
It comes from the execution of `fat-jar-kafka-010-example`, I just copied it. Will remove.

pnowojski commented on flink-examples/flink-examples-streaming/pom.xml:
```diff
@@ -568,6 +585,49 @@ under the License.
+								</filter>
+							</filters>
+							<artifactSet>
+								<excludes>
```
ditto: add `0.8` and `0.11`?

pnowojski commented on flink-examples/flink-examples-streaming/pom.xml:
```diff
@@ -568,6 +585,49 @@ under the License.
+									</includes>
+								</filter>
+							</filters>
+							<artifactSet>
```
nit: please reorder `artifactSet` so that both `KafkaExample` and `Kafka10Example` sections are consistent. (either move this to the top, or move `artifactSet` in `Kafka10Example` to the bottom).

pnowojski commented:
@yanghua I've created a jira ticket to fix `set -e` in other scripts https://issues.apache.org/jira/browse/FLINK-10711 (that's just FYI)

yanghua commented:
@pnowojski I have fixed some of the issues you mentioned. Regarding the shell error, I don't actually change the code. I didn't have to change this logic. I have seen it in my local tests, but not every time.

pnowojski commented:
@yanghua but it need to be fixed somehow, otherwise with `set -e` the script will be failing (from time to time?). On the other hand without `set -e` this check might be ignored.

yanghua commented:
Let me try to locate it. This code was written by @tzulitai  before.

yanghua commented:
@tzulitai shell function `get_partition_end_offset` in kafka-common.sh : 

```
$KAFKA_DIR/bin/kafka-consumer-groups.sh --describe --group dummy-consumer --bootstrap-server localhost:9092
```

always and just print : 

```
Consumer group 'dummy-consumer' has no active members.
```

Any suggestion?

cc @pnowojski

yanghua commented on flink-end-to-end-tests/test-scripts/kafka-common.sh:
```diff
@@ -120,14 +126,7 @@ function get_partition_end_offset {
-    | grep "$topic \+$partition" \
-    | tr -s " " \
-    | cut -d " " -f 4
+  $KAFKA_DIR/bin/kafka-run-class.sh kafka.tools.GetOffsetShell --broker-list localhost:9092 --topic $topic --partitions $partition --time -1 | cut -d ":" -f 3
```
I changed the way of getting the offset. @pnowojski

yanghua commented:
@pnowojski We cannot add `set -e` to top of the modified files. This action would cause the test to fail.

pnowojski commented:
How do they fail? I would say they should be fixed and if errors are expected, they should be handled.

yanghua commented:
@pnowojski You can check the [old Travis log](https://api.travis-ci.org/v3/job/448260257/log.txt). Just search : 

```
[FAIL] 'Kafka end-to-end test' failed after 0 minutes and 57 seconds! Test exited with exit code 1
```

The shell script does not report the error, but the result would show this information. If we do not add `set -e`, it would be OK.

pnowojski commented:
@yanghua it's not OK if we do not add `set -e`. The error would be still there, but it would just be silently ignored. 

You might want to to `set -x` as well for debugging, to print the commands that are being executed.

yanghua commented:
@pnowojski I have debugged locally, and the new changes have fixed the shell error. Before fixing, even if I do not add `set -e`, it will report an error. Now, I add `set-e', which still doesn't report any errors, but the result doesn't show 'pass'. When I remove the `set -e`, the result will show that the test is passed. Considering that there was no `set-e' in these scripts before, I think it was this sentence that caused the test to fail.

yanghua commented:
> shell function `get_partition_end_offset` in kafka-common.sh :
> 
> ```
> $KAFKA_DIR/bin/kafka-consumer-groups.sh --describe --group dummy-consumer --bootstrap-server localhost:9092
> ```
> always and just print :
> 
> ```
> Consumer group 'dummy-consumer' has no active members.
> ```

@pnowojski  This is the reason for the shell error. Now I changed the way of getting the offset, so it fixed.

yanghua commented:
@pnowojski I can give you more details : 

added `set -x` and `set -e`: 

```
+ stop_kafka_cluster
+ ${FLINK_DIR}/flink-end-to-end-tests/test-scripts/temp-test-directory-01N/kafka_2.11-2.0.0/bin/kafka-server-stop.sh
+ ${FLINK_DIR}/flink-end-to-end-tests/test-scripts/temp-test-directory-01N/kafka_2.11-2.0.0/bin/zookeeper-server-stop.sh
++ grep -i 'kafka\.Kafka'
++ jps -vl
++ grep java
++ grep -v grep
++ awk '{print $1}'
+ PIDS=55405
+ '[' '!' -z 55405 ']'
+ kill -s TERM 55405
++ jps -vl
++ grep java
++ grep -i QuorumPeerMain
++ grep -v grep
++ awk '{print $1}'
+ PIDS=

[FAIL] 'Modern Kafka end-to-end test' failed after 0 minutes and 57 seconds! Test exited with exit code 1

Stopping taskexecutor daemon (pid: 56949) on host ${MY_HOST}.
Stopping standalonesession daemon (pid: 56518) on host ${MY_HOST}.
No zookeeper daemon to stop on host ${MY_HOST}.
```

But, just add `set -x`

```
+ stop_kafka_cluster
+ ${FLINK_DIR}/flink-end-to-end-tests/test-scripts/temp-test-directory-21N/kafka_2.11-2.0.0/bin/kafka-server-stop.sh
+ ${FLINK_DIR}/flink-end-to-end-tests/test-scripts/temp-test-directory-21N/kafka_2.11-2.0.0/bin/zookeeper-server-stop.sh
++ jps -vl
++ grep -i 'kafka\.Kafka'
++ grep java
++ grep -v grep
++ awk '{print $1}'
+ PIDS=62800
+ '[' '!' -z 62800 ']'
+ kill -s TERM 62800
++ jps -vl
++ grep java
++ grep -i QuorumPeerMain
++ grep -v grep
++ awk '{print $1}'
+ PIDS=
+ '[' '!' -z '' ']'
+ mv -f ${FLINK_DIR}/flink-dist/target/flink-1.7-SNAPSHOT-bin/flink-1.7-SNAPSHOT/conf/flink-conf.yaml.bak ${FLINK_DIR}/flink-dist/target/flink-1.7-SNAPSHOT-bin/flink-1.7-SNAPSHOT/conf/flink-conf.yaml

[PASS] 'Modern Kafka end-to-end test' passed after 1 minutes and 0 seconds! Test exited with exit code 0.

Stopping taskexecutor daemon (pid: 64351) on host ${MY_HOST}.
Stopping standalonesession daemon (pid: 63917) on host ${MY_HOST}.
No zookeeper daemon to stop on host ${MY_HOST}.
Deleted all files under ${FLINK_DIR}/flink-dist/target/flink-1.7-SNAPSHOT-bin/flink-1.7-SNAPSHOT/log/
Deleted ${FLINK_DIR}/flink-end-to-end-tests/test-scripts/temp-test-directory-21N

[PASS] All tests passed
```

the problem : 

```shell

function stop_kafka_cluster {
  $KAFKA_DIR/bin/kafka-server-stop.sh
  $KAFKA_DIR/bin/zookeeper-server-stop.sh

  PIDS=$(jps -vl | grep -i 'kafka\.Kafka' | grep java | grep -v grep | awk '{print $1}')

  if [ ! -z "$PIDS" ]; then
    kill -s TERM $PIDS
  fi

  # here
  PIDS=$(jps -vl | grep java | grep -i QuorumPeerMain | grep -v grep | awk '{print $1}') 

  if [ ! -z "$PIDS" ]; then
    kill -s TERM $PIDS
  fi
}
```

pnowojski commented:
Please try adding `|| echo ""` after `awk '{print $1}'`, like in this example:
```
set -x
set -e
set -u
set -o pipefail

PIDS=$(jps -vl | grep java | grep -i QuorumPeerMain | grep -v grep | awk '{print $1}'|| echo "")
if [ ! -z "$PIDS" ]; then
  echo "killing" $PIDS
else
  echo "not killing" $PIDS
fi

echo "DONE"
```

yanghua commented:
@pnowojski Thank you for your guidance, now the test can pass. I submitted the last change, can you help me to review it again? Thank you.

pnowojski commented on flink-end-to-end-tests/test-scripts/kafka-common.sh:
```diff
@@ -80,13 +82,13 @@ function stop_kafka_cluster {
   $KAFKA_DIR/bin/zookeeper-server-stop.sh
 
-  PIDS=$(jps -vl | grep -i 'kafka\.Kafka' | grep java | grep -v grep | awk '{print $1}')
+  PIDS=$(jps -vl | grep -i 'kafka\.Kafka' | grep java | grep -v grep | awk '{print $1}'|| echo "")
```
Add comment `# Terminate Kafka process if it still exists`

pnowojski commented on flink-end-to-end-tests/test-scripts/kafka-common.sh:
```diff
@@ -80,13 +82,13 @@ function stop_kafka_cluster {
   fi
 
-  PIDS=$(jps -vl | grep java | grep -i QuorumPeerMain | grep -v grep | awk '{print $1}')
+  PIDS=$(jps -vl | grep java | grep -i QuorumPeerMain | grep -v grep | awk '{print $1}'|| echo "")
```
`# Terminate QuorumPeerMain process if it still exists`

pnowojski commented on flink-examples/flink-examples-streaming/pom.xml:
```diff
@@ -60,6 +60,18 @@ under the License.
 			<groupId>org.apache.flink</groupId>
 			<artifactId>flink-connector-kafka-0.10_${scala.binary.version}</artifactId>
 			<version>${project.version}</version>
+			<exclusions>
```
Why are you excluding this dependency here? end-to-end tests should be the final validation that our code works and having to modify dependencies doesn't sound good and might hide some error.

I wonder if maybe a better solution would be to create some separate sub modules for examples with kafka 0.10 and kafka 2.0?

@tzulitai what do you think?

tzulitai replied:
I also think that the Kafka 2.0 example should be in a separate module.
Take for example our Elasticsearch E2E tests: each version of the Elasticsearch job lives within its own module, so that we don't need to do any extra things with the dependency packaging.

On the other hand, as you can see, the Elasticsearch jobs live under the `flink-end-to-end-tests` module.
Perhaps it would make sense to do the same for Kafka, instead of relying on the released example jobs as we do right now?

yanghua replied:
@pnowojski I ruled it out because I think the 2.0 version of the connector relies on the kafka client. But then I chose to exclude different connectors for different packaging modes. I can try it out and undo this code to see if it will cause problems.

yanghua replied:
@pnowojski @tzulitai The result is not as expected (it failed), and even if we use `artifactSet / exculde` to exclude other connectors, the dependency on kafka-client is still wrong. So I'll take your advice and split the test implementation of the modern Kafka connector into separate modules. But because of the logical similarity, do you recommend that I introduce a base implementation (in a new base module)?

yanghua replied:
@pnowojski I have split the modern kafka connector into a separate module, but I have not restored the code for the flink-examples-streaming module. Please review again (test failure is irrelevant).

pnowojski commented on flink-end-to-end-tests/flink-kafka-test-base/src/main/java/org/apache/flink/kafka/tests/KafkaEvent.java:
```diff
@@ -0,0 +1,74 @@
+ * when keying a {@link org.apache.flink.streaming.api.datastream.DataStream} of such a type.
+ * For a demonstration of this, see the code in {@link Kafka010Example}.
+ */
+public class KafkaEvent {
```
I have talked with @tzulitai and we concluded that to avoid this duplication it would be best to:

- Move submodule (and rename) `flink-kafka-test-base` to `flink-streaming-examples/flink-kafka-example-base`
- Kafka version specifics (`flink-streaming-examples/flink-kafka-example` and `flink-streaming-examples/flink-kafka-0.10-example`) extend that base
- End to end tests just use the example jobs for testing

yanghua replied:
OK, it sounds a good solution.

yanghua replied:
@pnowojski I think currently there is not a module named `flink-streaming-examples` but a module named `flink-examples-streaming` which is not a parent module (and contains many kind of examples). Do you mean that I need to create a module called `flink-streaming-examples`?

pnowojski replied:
No, sorry about a typo. I was thinking about creating a submodule in the existing one `flink-examples/flink-examples-streaming/`

yanghua replied:
@pnowojski and @aljoscha It is hoped that the end-to-end testing for the modern Kafka connector can be merged before the 1.7.0 release.

pnowojski replied:
I think the new flat organisation that you proposed is fine, makes it more fine grained and easier to understand/re-use by the user.

yanghua commented on flink-end-to-end-tests/flink-kafka-test/pom.xml:
```diff
@@ -0,0 +1,91 @@
+
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-connector-kafka-base_${scala.binary.version}</artifactId>
```
I didn't introduce it at first, this was done to explicitly exclude the base module's dependency on the lower version of the kafka client. Otherwise it will cause a compilation error.

pnowojski replied:
What was the error? We should definitely look into it, since we shouldn't expect users to use the same hack in their poms.

yanghua replied:
```
03:59:30.102 [WARNING] 
Dependency convergence error for org.apache.kafka:kafka-clients:0.10.2.1 paths to dependency are:
+-org.apache.flink:flink-kafka-test:1.7-SNAPSHOT
  +-org.apache.flink:flink-connector-kafka_2.11:1.7-SNAPSHOT
    +-org.apache.flink:flink-connector-kafka-base_2.11:1.7-SNAPSHOT
      +-org.apache.kafka:kafka-clients:0.10.2.1
and
+-org.apache.flink:flink-kafka-test:1.7-SNAPSHOT
  +-org.apache.flink:flink-connector-kafka_2.11:1.7-SNAPSHOT
    +-org.apache.kafka:kafka-clients:2.0.0

03:59:30.103 [WARNING] Rule 0: org.apache.maven.plugins.enforcer.DependencyConvergence failed with message:
Failed while enforcing releasability. See above detailed error message.
```

details : https://api.travis-ci.org/v3/job/449173244/log.txt

yanghua replied:
OK, agree.

pnowojski commented on flink-examples/flink-examples-streaming-kafka-0.10/pom.xml:
```diff
@@ -0,0 +1,76 @@
+	<artifactId>flink-examples-streaming-kafka-0.10</artifactId>
+
+	<dependencies>
+		<dependency>
```
It will be better if you do not relay on transitive dependencies here and explicitly list `flink-streaming-java` and `flink-java` dependencies in the pom. Now it works because those modules are transitive dependency of `flink-connector-kafka-0.10` while `Kafka010Example` explicitly uses/depends on them.

pnowojski commented on flink-examples/flink-examples-streaming-kafka/pom.xml:
```diff
@@ -0,0 +1,78 @@
+	<name>flink-examples-streaming-kafka</name>
+
+	<dependencies>
+		<dependency>
```
ditto: add `flink-streaming-java` and `flink-java`?

yanghua commented:
@pnowojski updated (squashed commits for `pull --rebase`) and I think what do mean is `flink-streaming-java` and `flink-streaming-scala`, right?

pnowojski commented:
I meant `flink-java` because you are explicitly using `ParameterTool` class from that package. Aren't you? You are not using anything from `flink-streaming-scala`, so you shouldn't depend on it.

yanghua commented:
@pnowojski Oh... yes, I take it for granted that this is streaming's example and does not rely on batch API. It's my fault. I'll fix it right away.

yanghua commented:
updated.

shaoxuan-wang commented:
@pnowojski @yanghua    I got this test failure when I rebase the flink master to my own repository which is connected with travis-ci.org.  And I did not find the origin travis test result in this thread. Is it possible this is a flaky test?

pnowojski commented:
@yanghua if the tests are not stable for 2.0 connector, they should be fixed.

@shaoxuan-wang what test failure do you have in mind?

shaoxuan-wang commented:
> @yanghua if the tests are not stable for 2.0 connector, they should be fixed.
> 
> @shaoxuan-wang what test failure do you have in mind?

 @pnowojski .  I cannot remember for now. But I a pretty sure it is either "Modern Kafka end-to-end test" or "Kafka end-to-end test". I quickly checked gitlog and find this pr on top of it.  When I rerun the test it passed. I guess one of these test is not stable. I will report it next when I see the failure.
